node/core/approval-voting/src/criteria.rs

```diff
@@ -356,23 +449,103 @@ fn compute_relay_vrf_modulo_assignments(
 			// has been executed.
 			let cert = AssignmentCert {
 				kind: AssignmentCertKind::RelayVRFModulo { sample: rvm_sample },
-				vrf: approval_types::VrfSignature {
-					output: approval_types::VrfOutput(vrf_in_out.to_output()),
-					proof: approval_types::VrfProof(vrf_proof),
+				vrf: VrfSignature {
+					output: VrfOutput(vrf_in_out.to_output()),
+					proof: VrfProof(vrf_proof),
 				},
 			};
 
 			// All assignments of type RelayVRFModulo have tranche 0.
 			assignments.entry(core).or_insert(OurAssignment {
-				cert,
+				cert: cert.into(),
 				tranche: 0,
 				validator_index,
 				triggered: false,
+				assignment_bitfield: core.into(),
 			});
 		}
 	}
 }
 
+fn assigned_cores_transcript(core_bitfield: &CoreBitfield) -> Transcript {
+	let mut t = Transcript::new(approval_types::v2::ASSIGNED_CORE_CONTEXT);
+	core_bitfield.using_encoded(|s| t.append_message(b"cores", s));
+	t
+}
+
+fn compute_relay_vrf_modulo_assignments_v2(
+	assignments_key: &schnorrkel::Keypair,
+	validator_index: ValidatorIndex,
+	config: &Config,
+	relay_vrf_story: RelayVRFStory,
+	leaving_cores: Vec<(CandidateHash, CoreIndex)>,
+	assignments: &mut HashMap<CoreIndex, OurAssignment>,
+) {
+	let mut assigned_cores = Vec::new();
+	// for rvm_sample in 0..config.relay_vrf_modulo_samples {
+	let maybe_assignment = {
+		let assigned_cores = &mut assigned_cores;
+		assignments_key.vrf_sign_extra_after_check(
+			relay_vrf_modulo_transcript_v2(relay_vrf_story.clone()),
+			|vrf_in_out| {
+				*assigned_cores = relay_vrf_modulo_cores(
+					&vrf_in_out,
+					config.relay_vrf_modulo_samples,
+					config.n_cores,
+				)
+				.into_iter()
+				.filter(|core| {
+					leaving_cores.iter().map(|(_, core)| core).collect::<Vec<_>>().contains(&core)
```

Comment:
this is reallocating repeatedly in a fairly hot loop. we should reallocate once before the loop or do `iter().map(...).any(|x| x == core)` in iteration